File: 2022/02/rps.py
#!/usr/bin/env python3.11

from aoc2022 import *


input = """A Y
B X
C Z
"""
input: str = inp()

games = [  l.split(" ") for l in input.splitlines() ]

R = 0
P = 1
S = 2

# P > R
# R > S
# S > P

def score(opp, you):
    print(opp, you)
    you = ord(you) - ord('X')
    opp = ord(opp) - ord('A')
    score = 1 + you
    s = score
    # print(score)

    if opp == you:
        # draw
        score += 3
        # print(f"P:{s} D")
    elif (you == P and opp == R) or (you == R and opp == S) or (you == S and opp == P):
        score += 6
        # print(f"P:{s} W")
    else:
        # print(f"P:{s} L")
        pass

    print(score)
    print()

    return score

# p1_ans = sum(score(*g) for g in games)
# print(p1_ans)
# p1(p1_ans)

def pick(opp_m, result):
    print(opp_m, result)
    opp = ord(opp_m) - ord('A')
    res = ord(result) - ord('X')

    move = ""
    if res == 0:
        # Lose:
        t = {}
        t[R] = S
        t[P] = R
        t[S] = P

        move = t[opp]
    elif res == 1:
        # Draw:
        move = opp
    else:
        # Win:
        t = {}
        t[S] = R
        t[R] = P
        t[P] = S

        move = t[opp]

    return score(opp_m, chr(move + ord('X')))

p2_ans = sum(pick(*g) for g in games)
print(p2_ans)
p2(p2_ans)

# p1(max(elves))
# p2(sum(sorted(elves)[-3:]))

File: 2022/02/rps.rs
#!/usr/bin/env rustr

use std::cmp::Ordering;

use aoc::*;

#[derive(Copy, Clone, PartialEq, Eq)]
enum Move {
    Rock = 1,
    Paper = 2,
    Scissors = 3,
}

impl Move {
    // can model as a rotate..
    fn gt(&self) -> Self {
        use Move::*;
        match self {
            Rock => Paper,
            Paper => Scissors,
            Scissors => Rock,
        }
    }

    fn le(&self) -> Self {
        use Move::*;
        match self {
            Rock => Scissors,
            Paper => Rock,
            Scissors => Paper,
        }
    }
}

impl Ord for Move {
    fn cmp(&self, other: &Self) -> Ordering {
        self.partial_cmp(other).unwrap()
    }
}

impl PartialOrd for Move {
    fn partial_cmp(&self, other: &Move) -> Option<Ordering> {
        use Move::*;

        match (self, other) {
            (Paper, Rock) => Some(Ordering::Greater),
            (Rock, Scissors) => Some(Ordering::Greater),
            (Scissors, Paper) => Some(Ordering::Greater),
            (a, b) if a == b => Some(Ordering::Equal),
            (a, b) => b.partial_cmp(a).map(Ordering::reverse),
        }
    }
}

impl FromStr for Move {
    type Err = ();

    fn from_str(s: &str) -> Result<Self, ()> {
        let res = match s {
            "A" | "X" => Move::Rock,
            "B" | "Y" => Move::Paper,
            "C" | "Z" => Move::Scissors,
            _ => return Err(()),
        };

        Ok(res)
    }
}

#[derive(Copy, Clone, PartialEq, Eq)]
enum Outcome {
    Lose,
    Draw,
    Win,
}

impl FromStr for Outcome {
    type Err = ();
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        use Outcome::*;
        Ok(match s {
            "X" => Lose,
            "Y" => Draw,
            "Z" => Win,
            _ => return Err(()),
        })
    }
}

fn main() {
    let mut aoc = AdventOfCode::new(2022, 2);
    let input = aoc.get_input();
    let games = input.lines().map(|l| l.split_once(' ').unwrap());

    let score = |(opp, you): (Move, Move)| {
        (you as u64)
            + match you.cmp(&opp) {
                Ordering::Greater => 6,
                Ordering::Equal => 3,
                _ => 0,
            }
    };
    let p1: u64 = games
        .clone()
        .map(|(opp, you)| (opp.parse().unwrap(), you.parse().unwrap()))
        .map(score)
        .sum();
    aoc.submit_p1(p1).unwrap();

    let p2: u64 = games
        .clone()
        .map(|(opp, you)| (opp.parse().unwrap(), you.parse().unwrap()))
        .map(|(opp, out): (Move, Outcome)| {
            let you = match out {
                Outcome::Lose => opp.le(),
                Outcome::Draw => opp,
                Outcome::Win => opp.gt(),
            };

            (opp, you)
        })
        .map(score)
        .sum();
    aoc.submit_p2(p2).unwrap();
}
